### PR TITLE
MRG: Functionality to change scoring method of searchlight

### PR DESCRIPTION
**Pull request by kaichogami**
Enhancement as discussed in #3475. Aim to support
```
sl = SearchLight(LogisticRegression(), scoring=roc_auc_score)
score = sl.fit(X, y).score(X, y)

score = cross_val_score(SearchLight(LogisticRegression()), X, y, scoring='roc_auc')
```

### CODE REVIEW

File: mne/decoding/tests/test_search_light.py [hunk 1]
```diff
@@ -54,6 +56,15 @@ def test_searchlight():
...
     assert_true(score.dtype == float)
 
+    # change score method
+    sl = SearchLight(LogisticRegression(), scoring=roc_auc_score)
+    sl.fit(X, y)
+    score = sl.score(X, y)
+    assert_array_equal(score.shape, [n_time])
+    assert_true(score.dtype == float)
+    sl = SearchLight(LogisticRegression())
+    assert_array_equal(cross_val_score(sl, X, y, scoring='roc_auc'), score)
```
Comment (kaichogami) [inline mne/decoding/tests/test_search_light.py]: `cross_val_score` fails with an error.
```
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/utils/validation.py", line 561, in column_or_1d
    raise ValueError("bad input shape {0}".format(shape))
ValueError: bad input shape (17, 10)
```

File: mne/decoding/tests/test_search_light.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 from ..search_light import SearchLight, GeneralizationLight
 from .. import Vectorizer
 
+from sklearn.metrics import accuracy_score, roc_auc_score
```
Comment (kaichogami) [inline mne/decoding/tests/test_search_light.py]: Unncessary import, will remove it in next commit.

File: mne/decoding/search_light.py [hunk 3]
```diff
@@ -338,7 +349,10 @@ def _sl_score(estimators, X, y):
     """
     n_iter = X.shape[-1]
     for ii, est in enumerate(estimators):
-        _score = est.score(X[..., ii], y)
+        if scoring is not None:
+            _score = scoring(X[..., ii], y)
+        else:
+            _score = est.score(X[..., ii], y)
```
Comment (kingjr) [inline mne/decoding/search_light.py]: you can do 
```
scorer = est.score if scoring is None else scoring
_score = scorer(X[..., ii], y)
```

File: mne/decoding/search_light.py [hunk 4]
```diff
@@ -226,13 +235,15 @@ def score(self, X, y):
         if X.shape[-1] != len(self.estimators_):
             raise ValueError('The number of estimators does not match '
                              'X.shape[-1]')
+        if self.scoring is not None:
+            X = self.predict(X)
```
Comment (kingjr) [inline mne/decoding/search_light.py]: Ok, I think that works, but the var names is a bit confusing. Can you add 
```
# If scoring is None (default), the predictions are internally generated by estimator.score()
# Else, we must first get the predictions to pass them to ad-hoc scorer.
```

File: mne/decoding/search_light.py [hunk 5]
```diff
@@ -316,7 +327,7 @@ def _sl_init_pred(y_pred, X):
     return y_pred
 
 
-def _sl_score(estimators, X, y):
+def _sl_score(estimators, scoring, X, y):
     """Aux. function to score SearchLight in parallel.
```
Comment (kingjr) [inline mne/decoding/search_light.py]: scoring is missing in the doc: You can add there the lines I suggested above to explicit what is going on.

Formally, I think it's not perfect because that means the functions does very different things depending on the input (i.e. X becomes a y_pred), but I guess it's private so we'll deal with it.

**Comment by kingjr** [comment]
I *think* the current design won't work, notably because we need to change the type of prediction as a function of the scoring. e.g. if you pass a roc_auc_score, the prediction should be continuous (e.g. `predict_proba` or `decision_function`) and not discrete (this is probably the cause of your error). Check how this is handled in sklearn cross_val_score.

Additionally, could you add tests to make sure that this is working:
```
SearchLight(Ridge(), scoring=my_scorer())
cross_val_score(SearchLight(Ridge()), X, y, scoring=my_scorer())
```

And once it passes, update the GeneralizationLight too

File: mne/decoding/search_light.py [hunk 6]
```diff
@@ -338,7 +358,10 @@ def _sl_score(estimators, X, y):
     """
     n_iter = X.shape[-1]
     for ii, est in enumerate(estimators):
-        _score = est.score(X[..., ii], y)
+        if scoring is not None:
+            _score = scoring(est, X[..., ii], y)
```
Comment (kaichogami) [inline mne/decoding/search_light.py]: @kingjr I made it with the signature `score_func(estimator, y, y_pred)` by using `make_scorer` so as to take care of the prediction type which `make_scorer` handles internally. However `cross_val_score` doesn't still pass. Let me know if you have any idea about it.

Comment (kingjr) [inline mne/decoding/search_light.py]: I'm afraid I don't know. Do you manage to make `cross_val_score(base_estimator, scoring=scoring)` pass?

Comment (agramfort) [inline mne/decoding/search_light.py]: your y is 2d and roc_auc requires y to be 1d or 1 column. You need to look over the columns of y.

Comment (kingjr) [inline mne/decoding/search_light.py]: @agramfort this is a single/multi class scoring problem. Isn't there a way to handle thus by sklearn scorers?

Comment (kaichogami) [inline mne/decoding/search_light.py]: `_score = scoring(est, X[..., ii], y)`, shouldn't this fail as well in case when `scoring` parameter is changed to `roc_auc_score` in the test?

Comment (kaichogami) [inline mne/decoding/search_light.py]: Ignore my last comment. Going through the code, I think error occurs because X is not 2D before being passed to `decision_function`. I can see two solutions.
* Make the input to be a 2D matrix or at least convert `X` to 2D before passing to `decision_function` 
* Follow the other API @kingjr mentioned `cross_val_score(make_pipeline(SearchLight(Ridge(), scoring=my_scorer()), X, y)`

I hope I make sense..

Comment (agramfort) [inline mne/decoding/search_light.py]: A sklearn score returns a float per fold. Look maybe at multi_output scorers






On Thu, Aug 11, 2016 at 3:32 PM +0200, "Jean-Rémi KING" <notifications@github.com> wrote:












In mne/decoding/search_light.py:
> @@ -338,7 +358,10 @@ def _sl_score(estimators, X, y):
>      """
>      n_iter = X.shape[-1]
>      for ii, est in enumerate(estimators):
> -        _score = est.score(X[..., ii], y)
> +        if scoring is not None:
> +            _score = scoring(est, X[..., ii], y)



@agramfort this is a single/multi class scoring problem. Isn't there a way to handle thus by sklearn scorers?

**Comment by kaichogami** [comment]
@kingjr I am not sure what should be done now.

**Comment by agramfort** [comment]
@kaichogami make CIs happy?

**Comment by kaichogami** [comment]
@agramfort yes, however I don't know how to solve the issue with `cross_val_score(clf, X, y, scorer='roc_auc').

**Comment by agramfort** [comment]
you **cannot** obtain more than one value per fold with a sklearn scorer. If you have multiple targets you need a custom scorer that return more than a float

so

cross_val_score(clf, X, y, scorer='roc_auc')

simple **cannot** work as is.

**Comment by kingjr** [comment]
@agramfort I think there is a confusion between multiple issues:

1. SearchLight involves multiple scores, so `cross_val_score(SearchLight(), scoring='roc_auc')` is not viable with the current sklearn API. IMO scoring metrics need not output a float by principle. Confusion matrices are a form of scoring metrics. In any case this debate is not to be discussed here.

2. We want to be able to parametrize the SearchLight scoring e.g. `SearchLight(clf, scoring='roc_auc')`. However this involves
   i) internally adapting the `predict` function: if you want to score with an AUC, you need to ensure that the predict function outputs an array of floats (e.g. distance to hyperplan, proba etc) and not of integers.
   ii) identifying or handling 2 or multi class problem: i.e. `predict_proba` generates a `y_pred, of shape (n_samples, n_classes)`, which can therefore not be passed to the sklearn roc_auc scorer. This is an API problem, not an intrinsic one, because we could in principle identify that there are 2 classes and compute the AUC. In case of multiclass, we could compute each auc and return the average AUC. All this should be solved/able at the sklearn level IMO

**Comment by agramfort** [comment]
does using check_scoring function from sklearn help?

see also implementation of make_scorer function.

a sklearn scorer made from a string knows if it needs proba or decision
function or just predict.

**Comment by kingjr** [comment]
@kaichogami how are we doing over here?

**Comment by kaichogami** [comment]
@kingjr I haven't touched it after my last comment. I'll look over it again and let you know if I can solve it, in few days.

**Comment by kaichogami** [comment]
@kingjr as Alex mentioned we need to design custom scorer to make `cross_val_score(clf, X, y, scorer='custom_scorer')` work. Or we will have to use it the other way you mentioned i.e, `cross_val_score(make_pipeline(SearchLight(Ridge(), scoring=roc_auc_score), X, y)`. I think using the second option is more viable than creating many custom scorers. 
Also as we have used `make_scorer` it should know whether it requires proba or decision
function or just predict.
Let me know what you think.

**Comment by kingjr** [comment]
+1 for the second option

**Comment by kaichogami** [comment]
Apologies @kingjr, the above design wouldn't work as well. We are now returning one value per fold, however we still have an array of score, one float for each estimator and this [line](https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/cross_validation.py#L1608) does not allow score to be anything other than a float, which makes sense. Error output
```
Traceback (most recent call last):
  File "/home/kaichogami/codes/dev_mne/env/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/kaichogami/codes/dev_mne/mne-python/mne/utils.py", line 791, in dec
    return function(*args, **kwargs)
  File "/home/kaichogami/codes/dev_mne/mne-python/mne/decoding/tests/test_search_light.py", line 66, in test_searchlight
    assert_array_equal(cross_val_score(make_pipeline(sl), X, y), score)
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/cross_validation.py", line 1480, in cross_val_score
    for train, test in cv)
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/externals/joblib/parallel.py", line 800, in __call__
    while self.dispatch_one_batch(iterator):
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/externals/joblib/parallel.py", line 658, in dispatch_one_batch
    self._dispatch(tasks)
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/externals/joblib/parallel.py", line 566, in _dispatch
    job = ImmediateComputeBatch(batch)
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/externals/joblib/parallel.py", line 180, in __init__
    self.results = batch()
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/externals/joblib/parallel.py", line 72, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/cross_validation.py", line 1593, in _fit_and_score
    test_score = _score(estimator, X_test, y_test, scorer)
  File "/home/kaichogami/codes/dev_mne/scikit-learn/sklearn/cross_validation.py", line 1653, in _score
    % (str(score), type(score)))
ValueError: scoring must return a number, got [ 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.] (<type 'numpy.ndarray'>) instead.
```
The only solution I can think of is to take average of scores, which could be a parameter in constructor. This should be made true while using with `cross_val_score`.
`cross_val_score(SearchLight(Ridge(), scoring=roc_auc_score, average=True), X, y)

**Comment by kingjr** [comment]
I think we have to dissociate two problems:

1. parametrize the scoring within the SearchLight (independently of the cross val score) i.e; `SearchLight(LogisiticRegression(), scorer=roc_auc_score)`; this should be close to what you already wrote

2. Deal with multiclass AUC: this is outside this PR, and will be addressed in sklearn (https://github.com/scikit-learn/scikit-learn/issues/3298)

3. make SearchLight compatible with `cross_val_score`, which expects float scores, not array. I think that may be a broader issue though, as it's similar to i.e. get a confusion matrix in a cross_val_score.

I recommend to solve 1, and leave 2 and 3 for now.

**Comment by kaichogami** [comment]
@kingjr I have removed the `cross_val_score` example and only solved the case 1 as mentioned in your previous comment.
I am not sure what is causing the travis and appveyor error.

**Comment by kingjr** [comment]
Not sure what s going on with Travis. Can you repush with a forced set-upstream?

**Comment by kaichogami** [comment]
Pushing again didn't help. Perhaps @Eric89GXL has some idea?

File: mne/decoding/search_light.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 
 import numpy as np
 
+from sklearn.metrics import make_scorer
```
Comment (Eric89GXL) [inline mne/decoding/search_light.py]: The Travis error says:
```
Found un-nested sklearn import")
```
In other words, this needs to be nested

Comment (kaichogami) [inline mne/decoding/search_light.py]: @Eric89GXL Thanks for your help!

**Comment by kaichogami** [comment]
I added the import in the relevant methods rather than using global imports, which solved the nested import issue.

File: mne/decoding/search_light.py [hunk 8]
```diff
@@ -19,6 +19,9 @@ class SearchLight(BaseEstimator, TransformerMixin):
     ----------
     base_estimator : object
         The base estimator to iteratively fit on a subset of the dataset.
+    scoring : callable, defaults to None
```
Comment (kingjr) [inline mne/decoding/search_light.py]: can you make it accept strings as well, similar to sklearn `cross_val_score`

File: mne/decoding/search_light.py [hunk 9]
```diff
@@ -30,10 +33,16 @@ def __repr__(self):
...
 
-    def __init__(self, base_estimator, n_jobs=1):
+    def __init__(self, base_estimator, scoring=None, n_jobs=1):
         _check_estimator(base_estimator)
         self.base_estimator = base_estimator
         self.n_jobs = n_jobs
+
+        if not (scoring is None or hasattr(scoring, '__call__')):
+            raise ValueError("scoring must be None or a callable type")
+
```
Comment (kingjr) [inline mne/decoding/search_light.py]: remove extra line

File: mne/decoding/search_light.py [hunk 10]
```diff
@@ -222,17 +231,24 @@ def score(self, X, y):
...
         self._check_Xy(X)
         if X.shape[-1] != len(self.estimators_):
             raise ValueError('The number of estimators does not match '
                              'X.shape[-1]')
+        # If scoring is None (default), the predictions are internally
+        # generated by estimator.score(). Else, we must first get the
+        # predictions based on the scorer.
+        self.scoring = (make_scorer(self.scoring) if self.scoring is not None
+                        else self.scoring)
```
Comment (kingjr) [inline mne/decoding/search_light.py]: object parameters should not be change.

You could store the scorer in `self.scorer_`, but I think we need not store the scorer, WDYT @agramfort ?

File: mne/decoding/search_light.py [hunk 11]
```diff
@@ -379,6 +402,13 @@ class GeneralizationLight(SearchLight):
         The number of jobs to run in parallel for both `fit` and `predict`.
         If -1, then the number of jobs is set to the number of cores.
     """
+    def __init__(self, base_estimator, n_jobs=1):
+        _check_estimator(base_estimator)
+        self.base_estimator = base_estimator
+        self.n_jobs = n_jobs
+
+        if not isinstance(self.n_jobs, int):
+            raise ValueError('n_jobs must be int, got %s' % n_jobs)
```
Comment (kingjr) [inline mne/decoding/search_light.py]: Why did you add the init? GeneralizationLight inherits from Searchlight and should have the same init params no?

File: mne/decoding/tests/test_search_light.py [hunk 12]
```diff
@@ -55,6 +57,13 @@ def test_searchlight():
     assert_true(np.sum(np.abs(score)) != 0)
     assert_true(score.dtype == float)
 
+    # change score method
```
Comment (kingjr) [inline mne/decoding/tests/test_search_light.py]: add test where 
* `scoring='foo'`
* check that sl.scoring == default

File: mne/decoding/tests/test_search_light.py [hunk 13]
```diff
@@ -55,6 +57,13 @@ def test_searchlight():
...
+    sl = SearchLight(LogisticRegression(), scoring=roc_auc_score)
+    sl.fit(X, y)
+    score = sl.score(X, y)
```
Comment (kingjr) [inline mne/decoding/tests/test_search_light.py]: add explicit test that check the score 

`assert_equal(score[0], roc_auc_score(y, LogisticRegression().fit(X,y).predict_proba(X)[:, 1))`

File: mne/decoding/search_light.py (hunk 10 above)
Comment (kaichogami) [inline mne/decoding/search_light.py]: I am now making all `scorer` checks in `__init__` which also handles string.

File: mne/decoding/search_light.py (hunk 11 above)
Comment (kaichogami) [inline mne/decoding/search_light.py]: It would otherwise result in error for the doc string for the `scorer` parameter, as in GeneralizationLight we have not modified it to change `scorer`.

File: mne/decoding/search_light.py [hunk 14]
```diff
@@ -30,10 +33,24 @@ def __repr__(self):
             repr_str += ', fitted with %i estimators' % len(self.estimators_)
         return repr_str + '>'
 
-    def __init__(self, base_estimator, n_jobs=1):
+    def __init__(self, base_estimator, scoring=None, n_jobs=1):
+        from sklearn.metrics import make_scorer, get_scorer
```
Comment (kaichogami) [inline mne/decoding/search_light.py]: @kingjr Is it alright to add imports in `__init__` of class?

**Comment by kaichogami** [comment]
@kingjr please have a look at the changes.
The travis error looks unrelated to the PR changes.

File: mne/decoding/search_light.py (hunk 11 above)
Comment (kingjr) [inline mne/decoding/search_light.py]: You shoudlnt change this init, but also add a scoring parameter to GeneralizationLight

File: mne/decoding/search_light.py (hunk 14 above)
Comment (kingjr) [inline mne/decoding/search_light.py]: No see below

File: mne/decoding/search_light.py [hunk 15]
```diff
@@ -30,10 +33,24 @@ def __repr__(self):
...
+        # If scoring is None (default), the predictions are internally
+        # generated by estimator.score(). Else, we must first get the
+        # predictions based on the scorer.
+        if not isinstance(self.scoring, str):
+            self.scoring = (make_scorer(self.scoring) if self.scoring is
+                            not None else self.scoring)
+
+        elif self.scoring is not None:
+            self.scoring = get_scorer(self.scoring)
+
```
Comment (kingjr) [inline mne/decoding/search_light.py]: Don't change the param at init, this will break the cloning behavior.

get_scorer and make_scorer should be called at .score() not init

File: mne/decoding/search_light.py [hunk 16]
```diff
@@ -328,6 +345,10 @@ def _sl_score(estimators, X, y):
     X : array, shape (n_samples, nd_features, n_estimators)
         The target data. The feature dimension can be multidimensional e.g.
         X.shape = (n_samples, n_features_1, n_features_2, n_estimators)
+    scoring : callable or None
```
Comment (kingjr) [inline mne/decoding/search_light.py]: or string?

File: mne/decoding/search_light.py [hunk 17]
```diff
@@ -375,11 +399,13 @@ class GeneralizationLight(SearchLight):
     ----------
     base_estimator : object
         The base estimator to iteratively fit on a subset of the dataset.
+    scoring : callable, string, defaults to None
```
Comment (kingjr) [inline mne/decoding/search_light.py]: scoring : callable | string | None

**Comment by kingjr** [comment]
Do you want to give it a look @agramfort ?

**Comment by agramfort** [comment]
LGTM

@kingjr merge if you're happy

fix the docstrings in master if needed or better in your next "privatization" PR

**Comment by kingjr** [comment]
I'll fix the docstring in the "privatizing" PR ;)

**Comment by kingjr** [comment]
Thanks @kaichogami !

**Comment by kaichogami** [comment]
@kingjr You are welcome! What is "privatizing" PR that is being talked about?

**Comment by kingjr** [comment]
We re making a release next week, but some of the recent decoding contribs
may be too unstable at the moment. I will temporarily make these changes
private until next week's release and put them back afterwards to give us
another 6 months to ensure the stability of the API.
